lib/feature/one_hot_encoder.cc

```diff
@@ -0,0 +1,126 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fstream"
+#include "sys/stat.h"
+#include <dirent.h>
+#include <string>
+
+#include "clink/feature/one_hot_encoder.h"
+#include "nlohmann/json.hpp"
+#include "tfrt/support/error_util.h"
+
+using namespace clink;
+using namespace std;
+using namespace nlohmann;
+
+llvm::Expected<SparseVector>
+OneHotEncoderModel::transform(const int value, const int columnIndex) {
+  if (columnIndex >= model_data_.featuresizes_size()) {
+    return tfrt::MakeStringError("Column index out of range.");
+  }
+
+  int len = model_data_.featuresizes(columnIndex);
+  if (value >= len) {
+    return tfrt::MakeStringError("Value out of range.");
+  }
+  if (getDropLast()) {
+    len -= 1;
+  }
+
+  SparseVector vector(len);
+  if (value < len) {
+    vector.set(value, 1.0);
+  }
+  return vector;
+}
+
+void OneHotEncoderModel::setDropLast(const bool is_droplast) {
+  params_.is_droplast = is_droplast;
+}
+
+bool OneHotEncoderModel::getDropLast() { return params_.is_droplast; }
+
+llvm::Error OneHotEncoderModel::setModelData(const std::string model_data_str) {
+  OneHotEncoderModelDataProto model_data;
+
+  if (!model_data.ParseFromString(model_data_str)) {
+    return tfrt::MakeStringError("Failed to parse modeldata.");
+  }
+
+  for (int i = 0; i < model_data.featuresizes_size(); i++) {
+    if (model_data.featuresizes(i) <= 0) {
+      return tfrt::MakeStringError(
+          "Model data feature size value must be positive.");
+    }
+  }
+
+  model_data_ = std::move(model_data);
+
+  return llvm::Error::success();
+}
+
+/*
+ * Given a directory path, gets any one file in the directory.
+ *
+ * Flink ML saves model data in a file whose name is unknown to C++. This
+ * function helps to locat that file.
+ */
+std::string getRandomFileInDirectory(std::string path) {
+  struct dirent *entry;
+  struct stat st;
+  DIR *dir = opendir(path.c_str());
+
+  if (dir == NULL) {
+    return NULL;
+  }
+  while ((entry = readdir(dir)) != NULL) {
+    const string full_file_name = path + "/" + entry->d_name;
+    if (stat(full_file_name.c_str(), &st) == -1)
+      continue;
+    bool is_directory = (st.st_mode & S_IFDIR) != 0;
+    if (!is_directory) {
+      return std::string(entry->d_name);
+    }
+  }
+  closedir(dir);
+  return NULL;
+}
+
+llvm::Expected<OneHotEncoderModel>
+OneHotEncoderModel::load(const std::string path) {
+  OneHotEncoderModel model;
+
+  ifstream params_input(path + "/metadata");
+  json params;
+  params << params_input;
+  std::string is_droplast = params["paramMap"]["dropLast"].get<std::string>();
+  model.setDropLast(is_droplast.compare("false"));
+  params_input.close();
+
+  ifstream model_data_input(path + "/data/" +
+                            getRandomFileInDirectory(path + "/data"));
```

Comment:
Do we need to return error if `getRandomFileInDirectory(...)` returns `NULL`?

Reply:
OK. I'll add the check.